### PR TITLE
[SLP]Improve costs in computeExtractCost() to avoid crash after D158449.

### PR DESCRIPTION
**Pull request by alexey-bataev**
Need to consider the length of the original vector for extractelements,
not the length, matched number of the scalars. It fixes 2 issues: 1)
improves cost estimation; 2) Fixes crashes after D158449.

### CODE REVIEW

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 1]
```diff
@@ -7067,30 +7067,53 @@ class BoUpSLP::ShuffleCostEstimator : public BaseShuffleAnalysis {
...
+        continue;
+      auto *VecTy = cast<FixedVectorType>(EE->getVectorOperandType());
+      NumElts = std::max(NumElts, VecTy->getNumElements());
+    }
+    auto *VecTy = FixedVectorType::get(VL.front()->getType(), NumElts);
     unsigned NumOfParts = TTI.getNumberOfParts(VecTy);
-
-    if (ShuffleKind != TargetTransformInfo::SK_PermuteSingleSrc ||
-        !NumOfParts || VecTy->getNumElements() < NumOfParts)
+    if (!NumOfParts || NumElts < NumOfParts)
```
Comment (valerydmit) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:7080]: When the condition NumElts < NumOfParts is true (assuming NumElts > 0 and NumOfParts is non-zero)?

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 2]
```diff
@@ -7067,30 +7067,53 @@ class BoUpSLP::ShuffleCostEstimator : public BaseShuffleAnalysis {
...
+      return TTI.getShuffleCost(ShuffleKind, VecTy, Mask);
+    unsigned EltsPerVector = PowerOf2Ceil(divideCeil(NumElts, NumOfParts));
+    int ValNum = -1;
+    int ValIdx = -1;
+    if (ShuffleKind != TargetTransformInfo::SK_PermuteSingleSrc &&
+        !all_of(enumerate(Mask), [&](auto &&Arg) {
```
Comment (valerydmit) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:7086]: Please add a comment explaining intent of the code. Is it to  check that mask indices are within sub-registers [if not poison]?

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp (hunk 1 above)
Comment (valerydmit) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:7080]: Found the condition originally appeared with 
```
commit 8e5f3d04f269dbe791076e775f1d1a098cbada01
Author: Fangrui Song <i@maskray.me>
Date:   Fri Apr 2 11:13:51 2021 -0700

    [SLPVectorizer] Fix divide-by-zero after D99719

    Will add a test case later.
```
But it looks like there is still no test case that executes the condition. @MaskRay , any chance to find/add the test case?

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 3]
```diff
@@ -7067,30 +7067,56 @@ class BoUpSLP::ShuffleCostEstimator : public BaseShuffleAnalysis {
...
                                      TTI::ShuffleKind ShuffleKind) {
-    auto *VecTy = FixedVectorType::get(VL.front()->getType(), VL.size());
+    unsigned NumElts = 0;
+    for (Value *V : VL) {
+      auto *EE = dyn_cast<ExtractElementInst>(V);
+      if (!EE)
+        continue;
+      auto *VecTy = cast<FixedVectorType>(EE->getVectorOperandType());
+      NumElts = std::max(NumElts, VecTy->getNumElements());
+    }
```
Comment (RKSimon) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:7077]: assert / bail if NumElts == 0 after the loop?

**Comment by alexey-bataev** [comment]
Ping!

**Review by valerydmit — APPROVED** [review]
LGTM

**Pull request closed without merging**